ios/engine/KMEI/KeymanEngine/Classes/Keyboard/InputViewController.swift

```diff
@@ -347,18 +347,40 @@ open class InputViewController: UIInputViewController, KeymanWebDelegate {
       perform(#selector(self.enableInputClickSound), with: nil, afterDelay: 0.1)
     }
 
-    var hasDeletedSelection = false
+    var deleteSelection = false
 
     if let selected = textDocumentProxy.selectedText {
       if selected.count > 0 {
-        textDocumentProxy.deleteBackward()
-        hasDeletedSelection = true
+        deleteSelection = true
       }
     }
 
-    if numCharsToDelete <= 0 || hasDeletedSelection {
+    if deleteSelection && newText == "" {
+      /*
+        if deleteSelection && newText == "", we have a backspace on
+        selected text.  Sadly, .insertText("")... does nothing.  Why, Apple!?
+       
+        The one silver lining:  Apple makes it impossible for users to select text
+        in a way that splits character clusters.
+       
+        So, we can just insert something that won't combine, like a ZWNJ, and then delete it.
+      */
+      let beforeManipulation = textDocumentProxy.documentContextBeforeInput ?? ""
+
+      textDocumentProxy.insertText("\u{200c}")
+      textDocumentProxy.deleteBackward()
+      
+      let afterManipulation = textDocumentProxy.documentContextBeforeInput ?? ""
+      
+      // For good measure, a canary to signal if our selected-text backspace handling
+      // goes awry.
+      if beforeManipulation != afterManipulation {
+        os_log(.error, log: KeymanEngineLogger.engine, "Could not cleanly execute backspace for selected text")
+      }
+      sendContextUpdate()
+      return
+    } else if numCharsToDelete <= 0 || deleteSelection {
```

Comment:
This method was already long, and the new additions would be easier to follow and maintain if they were encapsulated in their own methods. If someone is reading this code and not concerned with the edge case, then it would be helpful for the edge case to be contained separately. Some of the comments could also be reduced if the new methods were named in a way that indicated what their purpose was.